8349755: Fix corner case issues in async UL

jdksjolen opened the pull request:
Hi,

There are a few corner-cases when logging with asynchronous UL, all of which I'd like to fix with this PR.

1. If the `AsyncLogWriter`  produces a log message,  then it'll deadlock as it'll wait for itself to release the `AsyncLogLock`.
2. If a producing thread produces another log message while enqueueing, then a deadlock will occur

We fix this by keeping a thread as the owner of the lock. If the lock is owned by the current thread, then we resort to synchronous logging.

Note, this fix introduces a third case:

3. If an unattached thread logs then there is no identity, so we cannot check if it holds the lock. We must therefore resort to synchronous logging in this case also

In UL, we have so far guaranteed Program-Order log message appearance. In other words, `log("A"); log("B");` is guaranteed to always appear in the order `AB` in the log output. For cases 1 and 3, we still keep this guarantee. However, for case 2 we break this guarantee. The recursive log message will appear before any unflushed log messages from the logging thread. As I see no simple way  of fixing this, along with the fact that this is highly unlikely to ever occur (this has never been observed in production), I've decided to break that guarantee and prioritize logs not being missing.

As it's preferable that we do not have recursive logging (as in case 2 and partially in case 1), I've decided to `assert` that these should not hold. In debug systems, during testing, we can find any accidental recursive logging. In production systems, we will handle this correctly by going back to synchronous logging.

Testing: GHA only, so far.

All the best,
Johan

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8349755](https://bugs.openjdk.org/browse/JDK-8349755): Fix corner case issues in async UL (**Bug** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23513/head:pull/23513` \
`$ git checkout pull/23513`

Update a local copy of the PR: \
`$ git checkout pull/23513` \
`$ git pull https://git.openjdk.org/jdk.git pull/23513/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23513`

View PR using the GUI difftool: \
`$ git pr show -t 23513`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23513.diff">https://git.openjdk.org/jdk/pull/23513.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23513#issuecomment-2649304389)
</details>

dholmes-ora commented on src/hotspot/share/logging/logAsyncWriter.cpp:
```diff
@@ -84,19 +91,43 @@ void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecora
+bool AsyncLogWriter::enqueue(LogFileStreamOutput& output, LogMessageBuffer::Iterator msg_iterator) {
+  if (resort_to_synchronous_logging()) {
+    return false;
+  }
```
```suggestion
  }
  
  // If we get here we know the AsyncLogWriter is initialized.
```

dholmes-ora reviewed (commented):
This looks good to me.  When you have nested logging I'm not even sure what the "right" order would be, so I'm not concerned in that regard.

Can we add some debug-only logging by the log producer so that we can actually test this?

Thanks

xmas92 commented on src/hotspot/share/logging/logAsyncWriter.cpp:
```diff
@@ -84,19 +91,43 @@ void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecora
+  AsyncLogWriter* alw = AsyncLogWriter::instance();
+  Thread* holding_thread = AsyncLogWriter::AsyncLogLocker::current_holder();
+  Thread* this_thread = Thread::current_or_null();
+  bool is_async_log_thread = this_thread != nullptr && alw == this_thread;
+  bool is_recursively_logging = this_thread != nullptr && holding_thread == this_thread;
+  assert(!is_recursively_logging, "Do not log while holding the Async log lock");
+  return is_async_log_thread    ||  // The async log producer is attempting to log, leading to recursive logging.
+         is_recursively_logging ||  // A thread enqueuing a message has attempted to log something.
+         alw == nullptr         ||  // There is no AsyncLogWriter instance yet.
+         this_thread == nullptr;    // The current thread is unattached.
```
I think an early return makes this more readable. Might even be worth just having four separate if statements.
```suggestion
  if (this_thread == nullptr) {
    // The current thread is unattached.
    return true;
  }
  bool is_async_log_thread = alw == this_thread;
  bool is_recursively_logging = holding_thread == this_thread;
  assert(!is_recursively_logging, "Do not log while holding the Async log lock");
  return is_async_log_thread    ||  // The async log producer is attempting to log, leading to recursive logging.
         is_recursively_logging ||  // A thread enqueuing a message has attempted to log something.
         alw == nullptr;            // There is no AsyncLogWriter instance yet.
```

```suggestion
  if (this_thread == nullptr) {
    // The current thread is unattached.
    return true;
  }
  
  if (holding_thread == this_thread) {
    // A thread, while enqueuing a message, has attempted to log something.
    // Do not log while holding the Async log lock.
    // Try to catch possible occurrences in debug builds.
    DEBUG_ONLY(ShouldNotReachHere();)
    return true;
  }
  
  if (alw == nullptr) {
    // There is no AsyncLogWriter instance yet.
    return true;
  }
  
  if (this_thread == alw) {
    // The async log producer is attempting to log, leading to recursive logging.
    return true;
  }
  
  return false;
```

xmas92 commented on src/hotspot/share/logging/logAsyncWriter.cpp:
```diff
@@ -84,19 +91,43 @@ void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecora
-void AsyncLogWriter::enqueue(LogFileStreamOutput& output, const LogDecorations& decorations, const char* msg) {
+// This function checks for cases where continuing with asynchronous logging may lead to stability issues, such as a deadlock.
+// If this returns true then we give up on logging asynchronously and do so synchronously instead.
+bool AsyncLogWriter::resort_to_synchronous_logging() {
```
The name `resort_to_synchronous_logging` says something about what the callers of `enqueue` should do. I think something more in the line of `is_enqueue_disallowed` / `is_async_logging_disallowed` would be better. Or probably inverting the the logic, and have it be `is_enqueue_allowed`.

xmas92 reviewed (changes requested):
Looks good. Have a few naming and code structure suggestions.

jdksjolen reviewed (commented):
Take requested changes

dholmes-ora commented on src/hotspot/share/logging/logAsyncWriter.cpp:
```diff
@@ -42,10 +42,16 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
 
   ~AsyncLogLocker() {
-    assert(_holder == Thread::current_or_null(), "must be");
+    assert(_holder == nullptr || _holder == Thread::current_or_null(), "must be");
```
Don't understand this change. We are in the destructor so if the holder is null then this must be an unattached thread. So the original assert suffices.

jdksjolen replied:
True, this is an unnecessary change.

dholmes-ora commented on test/hotspot/jtreg/runtime/logging/AsyncDeathTestDebug.java:
```diff
@@ -0,0 +1,46 @@
+public class AsyncDeathTestDebug {
+    public static void main(String[] args) throws Exception {
+        ProcessBuilder pb =
+            ProcessTools.createLimitedTestJavaProcessBuilder("-Xlog:async", "-Xlog:all=debug");
```
If we are going to crash we should add `-XX:-CreateCoredumpOnCrash`

dholmes-ora replied:
Also does this mean we are going to crash on the very first log_debug? You have two logging statements that will trigger the crash but only one of them can be reached with the way the testing is structured.

dholmes-ora replied:
When a test knows it will crash we don't want to generate a coredump as we don't care about it. Dumping can take a significant time on some systems leading to timeouts.

jdksjolen replied:
Thanks David, I missed the `-` and was under the pretension that we don't create coredumps by default :).

dholmes-ora commented on test/hotspot/jtreg/runtime/logging/AsyncDeathTestNonProduct.java:
```diff
@@ -0,0 +1,46 @@
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+
+public class AsyncDeathNonProduct {
```
Did you mean to call this `AsyncDeathNonDebug` ??

You could just have one test file and two @test sections: one for debug and one for non-debug, then adapt the logic accordingly.  But this non-debug version isn't verifying the correct execution of recursive logging in product mode, because there is no actual recursive logging being performed - this test isn't doing anything. You would need some product logging to actually verify that but then anything that did `-Xlog:all` would crash. So perhaps just delete this version of the test.

jdksjolen replied:
I meant to call it `NonProduct`, and the rest of your message is confusing to me so I've probably misunderstood what my test does and what `NOT_PRODUCT` means :-).

I've got:

```c++
NOT_PRODUCT(log_debug(deathtest)("Induce a recursive log for testing");)
```

I thought `NON_PRODUCT` means that this is included regardless of  `assert`s being enabled, but will not be shipped in a product build of the JDK. That's what I want to do, as that let's me test recursive logging without any asserts being triggered.

I'll look into the test section suggestion.

dholmes-ora replied:
I've misunderstood your intention, but we only run tests on debug and product and your `@requires !vm.debug` means the test will run on product and it will do nothing because the log statements will not be there.

jdksjolen replied:
Aha, I see. Well, I think that I can re-work the code that let's us test it to accommodate for that.

dholmes-ora commented on src/hotspot/share/logging/logAsyncWriter.cpp:
```diff
@@ -42,10 +42,16 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
   }
+
+  void wait() {
+    _holder = nullptr;
+    _instance->_lock.wait(0/* no timeout */);
+    _holder = Thread::current_or_null();
```
```suggestion
    Thread* saved_holder = _holder;
    _holder = nullptr;
    _instance->_lock.wait(0/* no timeout */);
    _holder = saved_holder;
```

jdksjolen commented:
Alright,

So I fixed the tests. The way I implemented the testing is a bit of a hack, but at least it's a local-to-UL hack.The checking function needs to know whether to crash or not and instead handle the message. In order to inform the function of this, we set a static bool variable to `true` earlier in the call stack.

dholmes-ora commented on src/hotspot/share/logging/logAsyncWriter.cpp:
```diff
@@ -31,18 +31,34 @@
   }
 
   ~AsyncLogLocker() {
+    assert(_holder == nullptr || _holder == Thread::current_or_null(), "must be");
```
The first clause is still redundant

dholmes-ora reviewed (approved):
This seems okay to me. The testing "hack" is not too awful - and better than what I had envisaged. :)

jdksjolen commented:
Thank you for your patience with this PR.

/integrate
